### ISSUE TITLE
Segmentation fault on Ubuntu 22.04 + Go 1.20.1

### ISSUE THREAD

**Issue by mrg0lden** [ebitengine/purego#101]
Hello,

I got a segmentation fault on WSL running Ubuntu 22.04 LTS when compiling the example with Go 1.20.1 using the latest available version `v0.2.0-alpha.1.0.20230213042757-086b819e2f9a`.

<details><summary>go env</summary>

```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/mo/.cache/go-build"
GOENV="/home/mo/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/mo/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/mo/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.20.1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="0"
GOMOD="/mnt/c/Users/MrGol/Projects/expermients/Go/purego/go.mod"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3406589635=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details><summary>The example code</summary>

```
package main

import (
	"fmt"
	"runtime"

	"github.com/ebitengine/purego"
)

func getSystemLibrary() string {
	switch runtime.GOOS {
	case "darwin":
		return "/usr/lib/libSystem.B.dylib"
	case "linux":
		return "libc.so.6"
	default:
		panic(fmt.Errorf("GOOS=%s is not supported", runtime.GOOS))
	}
}

func main() {
	libc := purego.Dlopen(getSystemLibrary(), purego.RTLD_NOW|purego.RTLD_GLOBAL)
	if err := purego.Dlerror(); err != "" {
		panic(err)
	}
	var puts func(string)
	purego.RegisterLibFunc(&puts, libc, "puts")
	puts("Calling C from Go without Cgo!")
}

```

</details>

Command used to compile
```
CGO_ENABLED=0 go build .
```

Output
```
Segmentation fault
```

<details><summary>gdb output</summary>

```
Starting program: /mnt/c/Users/MrGol/Projects/expermients/Go/purego/xpr-purego
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7e28540 in _int_malloc (av=av@entry=0x7ffff7f9dc80 <main_arena>, bytes=bytes@entry=640) at ./malloc/malloc.c:4375
4375    ./malloc/malloc.c: No such file or directory.
```

</details>

**Comment by hajimehoshi** [comment]
@TotallyGamerJet

**Comment by TotallyGamerJet** [comment]
I'll take a look!

**Comment by TotallyGamerJet** [comment]
Offending commit is #73

**Comment by klauspost** [comment]
@TotallyGamerJet I still see it using latest version:

```
(gdb) r
Starting program: /home/user1/cmd/s2c/s2c
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7e2a540 in _int_malloc (av=av@entry=0x7ffff7f9fc80 <main_arena>, bytes=bytes@entry=640) at ./malloc/malloc.c:4375
4375    ./malloc/malloc.c: No such file or directory.
(gdb) where
#0  0x00007ffff7e2a540 in _int_malloc (av=av@entry=0x7ffff7f9fc80 <main_arena>, bytes=bytes@entry=640)
    at ./malloc/malloc.c:4375
#1  0x00007ffff7e2aa49 in tcache_init () at ./malloc/malloc.c:3245
#2  0x00007ffff7e2b25e in tcache_init () at ./malloc/malloc.c:3241
#3  __GI___libc_malloc (bytes=64) at ./malloc/malloc.c:3306
#4  0x000000000065dee2 in github.com/ebitengine/purego/internal/fakecgo.call5 ()
    at e:/gopath/pkg/mod/github.com/ebitengine/purego@v0.4.0-alpha/internal/fakecgo/trampolines_linux_amd64.s:72
#5  0x000000000065d9ec in github.com/ebitengine/purego/internal/fakecgo.malloc (size=<optimized out>)
    at e:/gopath/pkg/mod/github.com/ebitengine/purego@v0.4.0-alpha/internal/fakecgo/symbols.go:25
#6  github.com/ebitengine/purego/internal/fakecgo.x_cgo_init (g=0x7ffff7f9fce0 <main_arena+96>, setg=6676224)
    at e:/gopath/pkg/mod/github.com/ebitengine/purego@v0.4.0-alpha/internal/fakecgo/go_linux_amd64.go:81
#7  0x000000000065e12a in github.com/ebitengine/purego/internal/fakecgo.x_cgo_init (g=0x8f5dc0 <runtime.g0>,
    setg=4615808) at <autogenerated>:1
#8  0x000000000065dd5c in x_cgo_init_trampoline ()
    at e:/gopath/pkg/mod/github.com/ebitengine/purego@v0.4.0-alpha/internal/fakecgo/trampolines_linux_amd64.s:29
#9  0x0000000000467a71 in runtime.rt0_go () at c:/go/src/runtime/asm_amd64.s:220
#10 0x0000000000000000 in ?? ()
(gdb)
```

Shouldn't this be fixed in 0.4.0-alpha?

Crosscompiled to linux/amd64 from:

```
λ go version
go version go1.20.1 windows/amd64
```

**Comment by TotallyGamerJet** [comment]
@klauspost It should be fixed. Perhaps what you are seeing is related to #106? Please open a new issue that links to this one and include full details of how to reproduce.
